plugins/inputs/openstack/openstack.go

```diff
@@ -0,0 +1,1059 @@
+	}
+	servers, err := servers.ExtractServers(page)
+	if err != nil {
+		return fmt.Errorf("unable to extract servers: %v", err)
+	}
+	for _, server := range servers {
+		o.servers[server.ID] = server
+		if o.ServerDiagnotics && server.Status == "ACTIVE" {
+			diagnostic, err := diagnostics.Get(o.compute, server.ID).Extract()
+			o.diag[server.ID] = diagnostic
+			if err != nil {
+				return fmt.Errorf("unable to get diagnostics for server(%v) %v", server.ID, err)
+			}
+		}
+	}
+	return nil
+}
+
+// gatherVolumes collects volumes from the OpenStack API.
+func (o *OpenStack) gatherVolumes() error {
+	//if !o.services.ContainsService(volumeV2Service) {
+	//	return nil
+	//}
+	page, err := volumes.List(o.volume, &volumes.ListOpts{AllTenants: true}).AllPages()
+	if err != nil {
+		return fmt.Errorf("unable to list volumes: %v", err)
+	}
+	v := []volume{}
+	volumes.ExtractVolumesInto(page, &v)
+	for _, volume := range v {
+		o.volumes[volume.ID] = volume
+	}
+	return nil
+}
+
+// gatherStoragePools collects storage pools from the OpenStack API.
+func (o *OpenStack) gatherStoragePools() error {
+	//if !o.services.ContainsService(volumeV2Service) {
+	//	return nil
+	//}
+	results, err := schedulerstats.List(o.volume, &schedulerstats.ListOpts{Detail: true}).AllPages()
+	if err != nil {
+		return fmt.Errorf("unable to list storage pools: %v", err)
+	}
+	storagePools, err := schedulerstats.ExtractStoragePools(results)
+	if err != nil {
+		return fmt.Errorf("unable to extract storage pools: %v", err)
+	}
+	for _, storagePool := range storagePools {
+		o.storagePools[storagePool.Name] = storagePool
+	}
+	return nil
+}
+
+// accumulateServerDiagnostics accumulates statistics from the compute(nova) service.
+func (o *OpenStack) accumulateServerDiagnostics(acc telegraf.Accumulator) {
+	for server_id, diagnostic := range o.diag {
+		tags := tagMap{
+			"server_id": server_id,
+		}
+		fields := fieldMap{
+			"server_id": server_id,
+		}
+		switch reflect.TypeOf(diagnostic).Kind() {
+		case reflect.Map:
+			s := reflect.ValueOf(diagnostic).MapRange()
+			port_map := make(map[string]string)
+			port_count := 0
+			for s.Next() {
+				if strings.Contains(s.Key().Interface().(string), "tap") {
+					tap_list := strings.Split(s.Key().Interface().(string), "_")
+					if val, ok := port_map[tap_list[0]]; ok {
+						fields["port_"+val+"_"+strings.Join(tap_list[1:], "_")] = s.Value().Interface().(float64)
+						tags["port_"+val] = tap_list[0]
+					} else {
+						port_count += 1
+						port_map[tap_list[0]] = strconv.Itoa(port_count)
+						fields["port_"+port_map[tap_list[0]]+"_"+strings.Join(tap_list[1:], "_")] = s.Value().Interface().(float64)
+						tags["port_"+port_map[tap_list[0]]] = tap_list[0]
+					}
+				} else {
+					fields[s.Key().Interface().(string)] = s.Value().Interface().(float64)
+				}
+			}
+			fields["no_of_ports"] = port_count
+		}
+		acc.AddFields("openstack_server_diagnostics", fields, tags)
```

Comment:
Do we really want different measurements here or would using `openstack` as measurement name with a `tag` `type="server diagnostics"` be better?

Reply:
That is a really good idea but might not be possible in this case as every type(Openstack services) has different keys(in terms of maps/json) or columns(in terms of DB table) it would not be possible to check whether if a particular column is applicable to a service or not (e.g. in case if the value is null/empty). Please let me know if you have a different approach towards this.

Reply:
Well the line-protocol supports these different "series" in the same measurement... But it might be difficult for other setups to deal with it. I rely on you in this point to do a sensible analysis given your use case. :-) Just asking to be sure you thought about it... In case you think it's better to have multiple measurements, please mark this conversation as "resolved".